server/src/main/scala/org/http4s/server/middleware/CSRF.scala

```diff
@@ -181,9 +185,9 @@ final class CSRF[F[_], G[_]] private[middleware] (
         c1 <- CSRF.cookieFromHeadersF[F, G](r, cookieName)
         raw1 <- F.fromEither(extractRaw(c1.content))
         raw2 <- F.fromEither(extractRaw(rawToken))
-        response <- if (CSRF.isEqual(raw1, raw2)) respAction
-        else F.raiseError[Response[G]](CSRFCheckFailed)
         newToken <- signToken[F](raw1) //Generate a new token to guard against BREACH.
+        response <- if (CSRF.isEqual(raw1, raw2)) http.run(embedInRequestAttributes(r, newToken))
+                    else F.raiseError[Response[G]](CSRFCheckFailed)
       } yield response.addCookie(ResponseCookie(name = cookieName, content = unlift(newToken))))
```

Comment:
Is there a reason why `createResponseCookie` is not used here, but used in `private[middleware] def validate` The difference is that `createResponseCookie` sets `httpOnly` to `true`, while default apply on `ResponseCookie` sets `httpOnly` to `false`. 

`TSecCSRF` where this middleware presumable originates from uses `ResponseCookie(name = cookieName, content = unlift(newToken)))` everywhere, i.e. no restriction on `httpOnly`.

@jmcardon Could you please elaborate a bit?

Reply:
```
  def createResponseCookie(token: CSRFToken): ResponseCookie =
    ResponseCookie(name = cookieName, content = token, httpOnly = true, secure = self.secure)
```

Is the client in case of `Ajax/XHR` requests not supposed to be able to get hold of tokens, or am I misunderstanding the usage of this middleware?

Reply:
re first point: I think that should also use `createResponseCookie` and you just caught a bug.

> Is the client in case of Ajax/XHR requests not supposed to be able to get hold of tokens, or am I misunderstanding something obvious?

They're not supposed to be able to get hold of tokens _in the cookie_. You have to explicitly embed the token to grab it in a route where you need to do this, but you're not supposed to arbitrarily be able to read it from the cookie as well.
The owasp page on double submit cookie pattern is pretty good about describing this.

Reply:
>You have to explicitly embed the token to grab it in a route where you need to do this, but you're not supposed to arbitrarily be able to read it from the cookie as well.

Can you perhaps show an example how this middleware is supposed to be used. I cannot see any set x-crsf-token header in the code